lib/usagereporter/teleport/aggregating/reporter.go

```diff
@@ -0,0 +1,316 @@
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package aggregating
+
+import (
+	"context"
+	"sync"
+	"time"
+
+	"github.com/gravitational/trace"
+	"github.com/sirupsen/logrus"
+
+	"github.com/gravitational/teleport/api/types"
+	prehogv1 "github.com/gravitational/teleport/gen/proto/go/prehog/v1alpha"
+	"github.com/gravitational/teleport/lib/backend"
+	usagereporter "github.com/gravitational/teleport/lib/usagereporter/teleport"
+	"github.com/gravitational/teleport/lib/utils"
+)
+
+const (
+	reportGranularity = 15 * time.Minute
+	reportTTL         = 60 * 24 * time.Hour
+)
+
+type ReporterConfig struct {
+	Backend backend.Backend
+	Log     logrus.FieldLogger
+
+	ClusterName    types.ClusterName
+	ReporterHostID string
+}
+
+func (cfg *ReporterConfig) CheckAndSetDefaults() error {
+	if cfg.Backend == nil {
+		return trace.BadParameter("missing Backend")
+	}
+	if cfg.Log == nil {
+		return trace.BadParameter("missing Log")
+	}
+	if cfg.ClusterName == nil {
+		return trace.BadParameter("missing ClusterName")
+	}
+	if cfg.ReporterHostID == "" {
+		return trace.BadParameter("missing ReporterHostID")
+	}
+	return nil
+}
+
+// NewReporter returns a new running [Reporter]. To avoid resource leaks,
+// GracefulStop must be called or the base context must be closed. The context
+// will be used for all backend operations.
+func NewReporter(
+	ctx context.Context,
+	cfg ReporterConfig,
+) (*Reporter, error) {
+	if err := cfg.CheckAndSetDefaults(); err != nil {
+		return nil, trace.Wrap(err)
+	}
+
+	anonymizer, err := utils.NewHMACAnonymizer(cfg.ClusterName.GetClusterID())
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+
+	baseCtx, baseCancel := context.WithCancel(ctx)
+	periodicCtx, periodicCancel := context.WithCancel(baseCtx)
+
+	r := &Reporter{
+		anonymizer: anonymizer,
+		svc:        reportService{cfg.Backend},
+		log:        cfg.Log,
+
+		clusterName:    anonymizer.AnonymizeNonEmpty(cfg.ClusterName.GetClusterName()),
+		reporterHostID: anonymizer.AnonymizeNonEmpty(cfg.ReporterHostID),
+
+		baseCtx:        baseCtx,
+		baseCancel:     baseCancel,
+		periodicCancel: periodicCancel,
+	}
+
+	r.wg.Add(1)
+	go r.runPeriodicFinalize(periodicCtx)
+
+	return r, nil
+}
+
+// Reporter aggregates and persists usage events to the backend.
+type Reporter struct {
+	anonymizer utils.Anonymizer
+	svc        reportService
+	log        logrus.FieldLogger
+
+	// clusterName is the anonymized cluster name.
+	clusterName []byte
+	// reporterHostID is the anonymized host ID of the reporter (this agent).
+	reporterHostID []byte
+
+	// wg tracks all long-lived or I/O bound goroutines, to be waited on during
+	// GracefulStop.
+	wg sync.WaitGroup
+	// baseCtx is the parent context for all operations (backend operations
+	// being the most important).
+	baseCtx context.Context
```

Comment:
How about a proper close API rather than stuffing multiple contexts in a struct like this?